src/MongoDB.Driver.Core/Core/NativeLibraryLoader/LibraryLoader.cs

```diff
@@ -0,0 +1,134 @@
+
+        public LibraryLoader(Func<SupportedPlatform, string> libraryLocator)
+        {
+            Ensure.IsNotNull(libraryLocator, nameof(libraryLocator));
+            ThrowIfNotSupportedPlatform();
+            _nativeLoader = CreateNativeLoader(libraryLocator);
+        }
+
+        // public methods
+        public T GetDelegate<T>(string name)
+        {
+            IntPtr ptr = _nativeLoader.GetFunctionPointer(name);
+            if (ptr == IntPtr.Zero)
+            {
+                throw new TypeLoadException($"The function {name} was not found.");
+            }
+
+            return Marshal.GetDelegateForFunctionPointer<T>(ptr);
+        }
+
+        // private methods
+        private INativeLibraryLoader CreateNativeLoader(Func<SupportedPlatform, string> libraryLocator)
+        {
+            var currentPlatform = GetCurrentPlatform();
+            var relativePath = libraryLocator(currentPlatform);
+            var absolutePath = GetAbsolutePath(relativePath);
+            return CreateNativeLoader(currentPlatform, absolutePath);
+        }
+
+        private INativeLibraryLoader CreateNativeLoader(SupportedPlatform currentPlatform, string libraryPath)
+        {
+            switch (currentPlatform)
+            {
+                case SupportedPlatform.Linux:
+                    return new LinuxLibraryLoader(libraryPath);
+                case SupportedPlatform.MacOS:
+                    return new DarwinLibraryLoader(libraryPath);
+                case SupportedPlatform.Windows:
+                    return new WindowsLibraryLoader(libraryPath);
+                default:
+                    throw new PlatformNotSupportedException($"Unexpected platform {currentPlatform}.");
+            }
+        }
+
+        private string FindLibraryOrThrow(string basePath, string relativePath)
+        {
+            var fullPath = Path.Combine(basePath, relativePath);
+            if (File.Exists(fullPath))
+            {
+                return fullPath;
+            }
+
+            throw new FileNotFoundException($"Could not find library {fullPath}.");
+        }
+
+        private string GetAbsolutePath(string relativePath)
+        {
+            var assembly = typeof(LibraryLoader).GetTypeInfo().Assembly;
+            var codeBase = assembly.CodeBase;
+            var uri = new Uri(codeBase);
+            var location = uri.AbsolutePath;
+            var basepath = Path.GetDirectoryName(location);
+
+            return FindLibraryOrThrow(basepath, relativePath);
+        }
+
+        private SupportedPlatform GetCurrentPlatform()
+        {
+#if NETSTANDARD1_5
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                return SupportedPlatform.MacOS;
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                return SupportedPlatform.Linux;
+            }
+            else
```

Comment:
This particular conditional compilation is weird.

How about we just remove the `else` clauses. Since each `then` clause returns the `else` is not needed.